test/gpu/vllm/vllm_test.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2024 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package vllm
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/docker/docker/api/types/mount"
+	"gvisor.dev/gvisor/pkg/test/dockerutil"
+	"gvisor.dev/gvisor/test/benchmarks/harness"
+)
+
+func TestMain(m *testing.M) {
+	harness.Init()
+	os.Exit(m.Run())
+}
+
+// BenchmarkVLLM runs a vLLM workload.
+func BenchmarkVLLM(b *testing.B) {
+	doVLLMTest(b)
+}
+
+func doVLLMTest(b *testing.B) {
+	serverMachine, err := harness.GetMachine()
+	if err != nil {
+		b.Fatalf("failed to get machine: %v", err)
+	}
+	defer serverMachine.CleanUp()
+
+	b.Run("opt-125", func(b *testing.B) {
+		ctx := context.Background()
+
+		b.ResetTimer()
+		b.StopTimer()
+
+		serverCtr := serverMachine.GetContainer(ctx, b)
+		defer serverCtr.CleanUp(ctx)
+		if err := harness.DropCaches(serverMachine); err != nil {
+			b.Skipf("failed to drop caches: %v. You probably need root.", err)
+		}
+
+		// Run vllm.
+		runOpts := dockerutil.GPURunOpts()
+		runOpts.Image = "gpu/vllm"
+		runOpts.Env = []string{"PYTHONPATH=$PYTHONPATH:/vllm"}
+
+		if err := serverCtr.Spawn(ctx, runOpts); err != nil {
+			b.Errorf("failed to run container: %v", err)
+		}
+		if out, err := serverCtr.WaitForOutput(ctx, "Uvicorn running on http://0.0.0.0:8000", 10*time.Minute); err != nil {
+			b.Fatalf("failed to start vllm model: %v %s", err, out)
+		}
+
+		for i := 1; i <= b.N; i++ {
+			func() {
+				// Increment in steps of 100.
+				// From experimentation metrics only starts to stabilize around a few hundreds of requests.
+				// Iterating until that point for every run seems wasteful.
+				requests := 100 * i
+
+				clientMachine, err := harness.GetMachine()
+				if err != nil {
+					b.Fatalf("failed to get machine: %v", err)
+				}
+				defer clientMachine.CleanUp()
+				clientCtr := clientMachine.GetNativeContainer(ctx, b)
```

Comment:
One note I'll make here... When using two containers on the same machine and one is "native" (`runc`) and the other is running in gVisor, the code paths that a packet takes to go from one to the other is much longer than the path that would be taken by a packet going from two native containers running on the same machine. This is because the Linux kernel's network stack optimizes `lo -> lo` traffic quite a bit, and with the gVisor network stack in the mix these optimizations don't apply.

Therefore, for benchmarks where networking is the bottleneck, if comparing unsandboxed vs gVisor performance, it's very important to put the server and the client on different machines for both the unsandboxed and the sandboxed case. Perhaps this benchmark isn't network-bound because the GPU is the bottleneck, but at some point as LLMs get more and more commoditized and optimized, I expect this to change.

There's no need to change anything for that in this benchmark, I am just pointing it out in case this is relevant to you when you run this benchmark.

Reply:
Thanks for pointing this out. I changed the client to a non-native container. I wasn't aware of this before and currently don't see a need for the native container.

Reply:
Note that using a "non-native" container in both cases doesn't really solve this problem, because when running this benchmark with the `runc` runtime, it'll still create `runc` containers for both client and server, so the `lo` optimization applies. (There is no way to fix this problem within this benchmark framework, because it is designed to run with Docker on a single machine, so again no action necessary here.)